### PR TITLE
Feat/actionable remediation ui behind ff

### PR DESCRIPTION
**Pull request by lili2311**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
- Render actionable remediation differently when present
- `--json` support is as before with remediation attached as is as an extra prop
- Project types that currently do not support remediation fall back to legacy format (so only npm, yarn, gradle, sbt, maven & ruby are using new format behind FF if FF is enabled)
If `remediation calculation` failed we fall back to old format
- TODO: tests

Blocked by https://github.com/snyk/snyk/pull/695
#### Screenshots
Screen Shot 2019-07-29 at 19 43 17
Screen Shot 2019-07-29 at 19 43 10

Screen Shot 2019-07-30 at 13 18 33

# --json
New prop on the json available:

```
"remediation": {
    "unresolved": [],
    "upgrade": {
      "rack@1.6.5": {
        "upgradeTo": "rack@1.6.11",
        "upgrades": [
          "rack@1.6.5"
        ],
        "vulns": [
          "SNYK-RUBY-RACK-72567"
        ]
      },
      "rack-cache@1.1": {
        "upgradeTo": "rack-cache@1.3.0",
        "upgrades": [
          "rack-cache@1.1",
          "rack-cache@1.1",
          "rack@1.6.5"
        ],
        "vulns": [
          "snyk:lic:rubygems:rack-cache:Unknown",
          "SNYK-RUBY-RACKCACHE-20031",
          "SNYK-RUBY-RACK-72567"
        ]
      },
      "rack-protection@1.5.3": {
        "upgradeTo": "rack-protection@2.0.0",
        "upgrades": [
          "rack-protection@1.5.3",
          "rack-protection@1.5.3",
          "rack-protection@1.5.3",
          "rack@1.6.5"
        ],
        "vulns": [
          "SNYK-RUBY-RACKPROTECTION-20394",
          "SNYK-RUBY-RACKPROTECTION-20395",
          "SNYK-RUBY-RACKPROTECTION-22019",
          "SNYK-RUBY-RACK-72567"
        ]
      }
    },
    "patch": {},
    "ignore": {}
  },
```

### CODE REVIEW

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          id,
+          basicVulnInfo[id].title, basicVulnInfo[id].severity, basicVulnInfo[id].isNew))
+      .join('\n');
+    upgradeTextArray.push(upgradeText + thisUpgradeFixes);
+  }
+  return upgradeTextArray;
+}
+
+function constructUnfixableText(unresolved: IssueData[]) {
+  const unfixableIssuesTextArray = [chalk.bold.white('Non-fixable issues:')];
```
Comment (lwywoo) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: Non-fixable sounds a bit weird, can we have something more like `There is no remediation available for these issues:`

File: src/lib/snyk-test/legacy.ts [hunk 2]
```diff
@@ -13,9 +14,46 @@ interface Patch {
...
+  LOW = 'low',
+  MEDIUM = 'medium',
+  HIGH = 'high',
+}
+
+export interface VulnMetaData {
+  id: string;
+  title: string;
+  description: string;
+  type: 'license' | 'vuln';
```
Comment (lwywoo) [inline src/lib/snyk-test/legacy.ts]: Why can the type for `VulnMetaData` be a license? Don't licences have their own metadata?

File: test/acceptance/display-test-results.test.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import * as tap from 'tap';
+// import * as path from 'path';
```
Comment (lwywoo) [inline test/acceptance/display-test-results.test.ts]: Do we need these commented out imports?

File: src/lib/snyk-test/legacy.ts (hunk 2 above)
Comment (lili2311) [reply src/lib/snyk-test/legacy.ts]: we treat a license as a vulnerability, so the vulns come in type `vuln` and `license`

File: test/acceptance/display-test-results.test.ts (hunk 3 above)
Comment (lili2311) [reply test/acceptance/display-test-results.test.ts]: nope

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts (hunk 1 above)
Comment (lili2311) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: Let's chat in the slack, I am following the designs here: https://snyk.slack.com/archives/CKLKY2ZRA/p1564391407005800

File: src/cli/commands/test/index.ts [hunk 4]
```diff
@@ -334,7 +340,8 @@ function displayResult(res, options: Options & TestOptions) {
     groupedVulnInfoOutput.join('\n\n') + '\n\n' +
     groupedDockerBinariesVulnInfoOutput.join('\n\n') + '\n\n' + meta + summary;
 
-  return prefix + body + multiProjAdvice + dockerAdvice + dockerSuggestion;
+  const ignoredIssues = '';
```
Comment (miiila) [inline src/cli/commands/test/index.ts]: Is `ignoredIssues` really needed if we immediately return it?

File: test/acceptance/display-test-results.test.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import * as tap from 'tap';
+import * as sinon from 'sinon';
+import * as _ from 'lodash';
+// tslint:disable-next-line:no-var-requires
+const snykTest = require('../../src/cli/commands/test');
+import * as snyk from '../../src/lib';
+
+const {test} = tap;
+(tap as any).runOnly = false; // <- for debug. set to true, and replace a test to only(..)
```
Comment (miiila) [inline test/acceptance/display-test-results.test.ts]: 😍

File: test/acceptance/display-test-results.test.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import * as tap from 'tap';
+import * as sinon from 'sinon';
+import * as _ from 'lodash';
+// tslint:disable-next-line:no-var-requires
+const snykTest = require('../../src/cli/commands/test');
```
Comment (miiila) [inline test/acceptance/display-test-results.test.ts]: Can't we move to import?

File: test/acceptance/display-test-results.test.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+(tap as any).runOnly = false; // <- for debug. set to true, and replace a test to only(..)
+
+const port = process.env.PORT = process.env.SNYK_PORT = '12345';
+process.env.SNYK_API = 'http://localhost:' + port + '/api/v1';
+process.env.SNYK_HOST = 'http://localhost:' + port;
+process.env.LOG_LEVEL = '0';
+
+test('`test ruby-app` meta when no vulns',  async (t) => {
+  chdirWorkspaces();
+  const stubbedResponse = require('./workspaces/ruby-app/test-graph-response-with-remediation.json');
```
Comment (miiila) [inline test/acceptance/display-test-results.test.ts]: `fs.readFileSync(JSON.parse('filePath')` please :-)

File: test/acceptance/display-test-results.test.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import * as _ from 'lodash';
+// tslint:disable-next-line:no-var-requires
+const snykTest = require('../../src/cli/commands/test');
+import * as snyk from '../../src/lib';
+
+const {test} = tap;
+(tap as any).runOnly = false; // <- for debug. set to true, and replace a test to only(..)
+
+const port = process.env.PORT = process.env.SNYK_PORT = '12345';
+process.env.SNYK_API = 'http://localhost:' + port + '/api/v1';
```
Comment (miiila) [inline test/acceptance/display-test-results.test.ts]: Why is this needed if we are stubbing `test`?

File: test/acceptance/display-test-results.test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    const res = error.message;
+    t.match(res, 'Upgrade rack@1.6.5 to rack@1.6.11 to fix', 'upgrade advice displayed');
+    t.match(res, 'Tested 3 dependencies for known issues, found 6 issues, 8 vulnerable paths.');
+  }
+
+  snykTestStub.restore();
+  t.end();
+});
+
+function chdirWorkspaces(subdir: string = '') {
```
Comment (miiila) [inline test/acceptance/display-test-results.test.ts]: Looks like this function is not used.

File: test/acceptance/display-test-results.test.ts (hunk 6 above)
Comment (lili2311) [reply test/acceptance/display-test-results.test.ts]: it is not compatible with how it's exported

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      isNew: vuln.isNew,
+      name: vuln.name,
+      version: vuln.version,
+      fixedIn: vuln.fixedIn,
+    }) as BasicVulnInfo;
+  });
+  const results = [chalk.bold.white('Remediation advice')];
+
+  const upgradeTextArray = contructUpgradesText(remediationInfo.upgrade, basicVulnInfo);
+  if (upgradeTextArray.length > 1) {
```
[the remark this answers is not in the document]
Comment (lili2311) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: first items is always the title of the section, let me refactor this to make it simpler

File: test/acceptance/display-test-results.test.ts (hunk 9 above)
Comment (lili2311) [reply test/acceptance/display-test-results.test.ts]: `chdirWorkspaces` is used as the first line in the test to change directory
